### PR TITLE
LPS-127613 field geolocation google maps

### PR DESCRIPTION
**Pull request by boton**
Issue: https://issues.liferay.com/browse/LPS-127613

---

**What's going on?**

In [useGeolocation](https://github.com/liferay/liferay-portal/blob/4acc81de59b202db303fd9c54b136694ceb23523/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/Geolocation/useGeolocation.es.js#L15-L16) we are importing the two maps providers and only the last imported provider is well constructed in our [global MapBase](https://github.com/liferay/liferay-portal/blob/master/modules/apps/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js#L836). 

When we import `MapGoogleMaps` the [MapBase implementations are updated](https://github.com/liferay/liferay-portal/blob/32f72ae28ccb9ba240d9a7902bde08d7e6854cd9/modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js#L97-L105).

When is loaded our `Liferay.MapBase.*Impl` are pointing to the [MapGoogleMaps](https://github.com/liferay/liferay-portal/blob/7c437a02cbc5f6a77c798076691961b702b2e21f/modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js#L115-L123) implementations.

Later we import `MapOpenStreetMap` and `Liferay.MapBase.*Impl` are updated another time leaving the `Liferay.MapBase.*Impl` pointing to [MapOpenStreetMap](https://github.com/liferay/liferay-portal/blob/master/modules/apps/map/map-openstreetmap/src/main/resources/META-INF/resources/js/MapOpenStreetMap.es.js#L119-L127) implementations and overwetting the [MapGoogleMaps](https://github.com/liferay/liferay-portal/blob/7c437a02cbc5f6a77c798076691961b702b2e21f/modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js#L115-L123).

When we try to use the `MapGoogleMaps` provider fails because the MapBase implementations are pointing to `MapOpenStreetMap`.

---

**The fix**

I move the global MapBase implementations initialization to the constructors. Instead of relying on the order of execution, we will only update MapBase implementations when a new `Map*` provider is created.

### CODE REVIEW

**Comment by boton** [comment]
ci:test:sf

**Comment by boton** [comment]
Please @liferay-frontend @p2kmgcl, Can you please review that I'm not doing something crazy here?

**Comment by p2kmgcl** [comment]
Looks good to me. Thanks so much @boton :smile:

File: modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js [hunk 1]
```diff
@@ -34,6 +34,12 @@ class MapGoogleMaps extends MapBase {
 	constructor(...args) {
 		super(...args);
 
+		MapBase.DialogImpl = GoogleMapsDialog;
```
Comment (julien) [inline modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js:37]: Why are we doing this in the constructor? Seems a bit strange.

File: modules/apps/map/map-openstreetmap/src/main/resources/META-INF/resources/js/MapOpenStreetMap.es.js [hunk 2]
```diff
@@ -34,6 +34,12 @@ class MapOpenStreetMap extends MapBase {
 	constructor(...args) {
 		super(...args);
 
+		MapBase.DialogImpl = OpenStreetMapDialog;
```
Comment (julien) [inline modules/apps/map/map-openstreetmap/src/main/resources/META-INF/resources/js/MapOpenStreetMap.es.js:37]: Same here

File: modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js (hunk 1 above)
Comment (boton) [reply modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js:37]: To update the `MapBase.*Impl` to the current map provider (`MapOpenStreetMap` or `MapGoogleMaps`) when [is created](https://github.com/liferay/liferay-portal/blob/ea44893298b4a31ba1da3206f5a0650e0d124cad/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/Geolocation/useGeolocation.es.js#L137-L162).

If we leave it outside the `MapBase.*Impl` are updated when we import it and the last provider imported overwrite the global MapBase no matter if it is used or not. 

Only the last imported module works if we leave it as is. And maybe this work in the past when we only use require one of the implementations like in [@liferay/map-openstreetmap](https://github.com/liferay/liferay-portal/blob/6a38677ec9534829a782d9b1a0961dba4d138dab/modules/apps/map/map-openstreetmap/src/main/resources/META-INF/resources/view.jsp#L38) ([require](https://github.com/liferay/liferay-portal/blob/master/modules/apps/map/map-openstreetmap/src/main/java/com/liferay/map/openstreetmap/internal/OpenStreetMapMapProvider.java#L78)) the same with `@liferay/map-google-maps` in `.jsp`.

Maybe we can move to an `initglobalImpl` method if you mean that 🤔

Comment (julien) [reply modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js:37]: @boton thanks for the information. Now that I've had a closer look at all these files, I understand what you mean. It still feels "strange" and maybe we could have solved that by using composition and not inheritance but, at this point if it works and the echo team is happy, feel free to forward. Thanks.

**Comment by boton** [comment]
ci:forward

File: modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js (hunk 1 above)
Comment (boton) [reply modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js:37]: @julien you are right, it feels a little bit strange and took me time to understand how it is working :D

Comment (julien) [reply modules/apps/map/map-google-maps/src/main/resources/META-INF/resources/js/MapGoogleMaps.es.js:37]: gotta love the magic of JavaScript :)

**Pull request closed without merging**